src/torch_ucc.cpp

```diff
@@ -718,6 +743,112 @@ ProcessGroupUCC::~ProcessGroupUCC() {
   }
 }
 
+#ifdef USE_CUDA
+// Return CUDA device with ordinal given by input rank.
+at::Device getCUDADeviceForRank(int rank) {
+  TORCH_CHECK(rank >= 0, "Invalid rank ", rank);
+  auto numGPUs = at::cuda::getNumGPUs();
+  int16_t deviceIdx = static_cast<int16_t>(rank % numGPUs);
+  return at::Device(at::DeviceType::CUDA, deviceIdx);
+}
+#endif
+
+void ProcessGroupUCC::runHealthCheck() {
+  // Run health check in a separate thread and wait on CV to handle timeouts.
+  // This design allows us to handle hangs.
+
+  // When size_ is 1, there is no need to do any communication at all.
+  if (size_ == 1) return;
+
+  struct HealthCheckData {
+    std::mutex healthCheckMutex;
+    std::condition_variable healthCheckCv;
+    bool ucxHealthCheckSuccess = false;
+    bool uccHealthCheckSuccess = false;
+    std::exception_ptr healthCheckException;
+  } healthCheckData;
+
+  auto t = std::thread([&healthCheckData, this]() {
+    std::list<c10::Device> devices{c10::kCPU};
+#ifdef USE_CUDA
+    if (at::cuda::is_available()) {
+      devices.emplace_front(getCUDADeviceForRank(rank_));
+    }
+#endif
+    for (auto device : devices) {
+      bool is_last_device = (device == c10::kCPU);
+      try {
+        auto oob = std::make_shared<torch_ucc_oob_coll_info_t>();
+        oob->rank = this->oob->rank;
+        oob->size = this->oob->size;
+        oob->store = this->oob->store;
+
+        std::vector<ucp_ep_h> eps;
+        ucc_team_h team = nullptr;
+        uint32_t comm_id;
+
+        auto comm = CommPG::get_comm(comm_id, device, oob, logger, true);
+        comm->ucx_connect_eps(eps, oob);
+        comm->ucx_disconnect_eps(eps, oob);
+        TORCH_UCC_LOG_INFO(TORCH_UCC_HEALTH_CHECK, "UCX library health check succeed.");
```

Comment:
nit: would be good to log which device we are checking.

```suggestion
        TORCH_UCC_LOG_INFO(
            TORCH_UCC_HEALTH_CHECK,
            c10::str(
                "UCX library health check succeed for device ",
                c10::DeviceTypeName(device))
        );
```